java/arcs/android/storage/handle/HandleFactory.kt

```diff
@@ -0,0 +1,144 @@
+ *
+ * Handles that are used by end-users will deal with [RawEntity], so this helper only bothers to
+ * create those types.
+ *
+ * It will create a [StorageProxy] for each new storage key, and keep a reference to it for as
+ * long as the [HandleFactory] exists.
+ *
+ * Instantiate it with the context and lifecycle that should own the resulting activate stores.
+*/
+@UseExperimental(kotlinx.coroutines.ExperimentalCoroutinesApi::class)
+class HandleFactory(
+    private val context: Context,
+    private val lifecycle: Lifecycle,
+    private val coroutineContext: CoroutineContext = EmptyCoroutineContext,
+    private val connectionFactory: ConnectionFactory? = null
+) {
+
+    companion object {
+        /**
+         * Convenience for making a ramdisk-backed reference mode key
+         */
+        fun ramdiskStorageKeyForName(name: String) = ReferenceModeStorageKey(
+            backingKey = RamDiskStorageKey("$name-backing"),
+            storageKey = RamDiskStorageKey("$name-storage")
+        )
+    }
+
+    private val singletonProxies = mutableMapOf<StorageKey, SingletonProxy<RawEntity>>()
+    private val setProxies = mutableMapOf<StorageKey, SetProxy<RawEntity>>()
+
+    /**
+     * Create a new SingletonHandle backed by an Android [ServiceStore]
+     */
+    suspend fun singletonHandle(
+        storageKey: StorageKey,
+        schema: Schema,
+        callbacks: Callbacks<SingletonOp<RawEntity>>? = null
+    ): SingletonHandle<RawEntity> {
+        val storeOptions = SingletonStoreOptions<RawEntity>(
+            storageKey = storageKey,
+            existenceCriteria = ExistenceCriteria.MayExist,
+            type = SingletonType(EntityType(schema)),
+            mode = StorageMode.ReferenceMode
+        )
+
+        val serviceStoreFactory = SingletonStoreFactory<RawEntity>(
+            context,
+            lifecycle,
+            ParcelableCrdtType.Singleton,
+            coroutineContext + Dispatchers.IO,
+            connectionFactory
+        )
+
+        val storageProxy = singletonProxies.getOrElse(storageKey) {
+            SingletonProxy(Store(storeOptions).activate(serviceStoreFactory), CrdtSingleton())
+        }
+
+        return SingletonHandle(storageKey.toKeyString(), storageProxy).also {
+            storageProxy.registerHandle(it)
+            it.callback = callbacks
+        }
+    }
+
+    /**
+     * Create a new [SetHandle] backed by an Android [ServiceStore]
+     */
+    suspend fun setHandle(
+        storageKey: StorageKey,
+        schema: Schema,
+        callbacks: Callbacks<SetOp<RawEntity>>? = null
+    ): SetHandle<RawEntity> {
+        val storeOptions = SetStoreOptions<RawEntity>(
+            storageKey = storageKey,
+            existenceCriteria = ExistenceCriteria.MayExist,
+            type = CollectionType(EntityType(schema)),
+            mode = StorageMode.ReferenceMode
+        )
+
+        val serviceStoreFactory = SetStoreFactory<RawEntity>(
+            context,
```

Comment:
Why not move most of this code into arcs.core.storage.handle as a base class and use the default store factory.  Create a method to return the platform specific storeFactory, and then in the android HandleFactory, override it with the ServiceStore version?

That way, the ArcHost code  (most of which resides in arcs.core) would share this too.

Reply:
+1, that also follows the goal of keeping relevant code in core so sdk is only platform specifics.

Reply:
Sounds good!

Reply:
(Giving it a shot for this PR)